src/bob/resource/internals.clj

```diff
@@ -0,0 +1,85 @@
+;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;   GNU General Public License for more details.
+;
+;   You should have received a copy of the GNU General Public License
+;   along with Bob. If not, see <http://www.gnu.org/licenses/>.
+
+(ns bob.resource.internals
+  (:require [aleph.http :as http]
+            [failjure.core :as f]
+            [bob.execution.internals :refer [pull build copy-dir!]]
+            [bob.util :refer [unsafe!]])
+  (:import (java.util.zip ZipInputStream)
+           (java.io File)
+           (java.nio.file Files)
+           (java.nio.file.attribute FileAttribute)))
+
+(defn- rm-r!
+  "Recursively deletes a directory."
+  [dir & [silently]]
+  (unsafe!
+    (letfn [(delete-f [^File file]
+              (when (.isDirectory file)
+                (doseq [child-file (.listFiles file)]
+                  (delete-f child-file)))
+              (clojure.java.io/delete-file file silently))]
+      (delete-f (clojure.java.io/file dir)))))
+
+(defn- extract-zip!
+  "Takes a java.util.zip.ZipInputStream `zip-stream` and extracts the content to the `out-dir`."
+  [^ZipInputStream zip-stream out-dir]
+  (unsafe!
+    (with-open [stream zip-stream]
+      (loop [entry (.getNextEntry stream)]
+        (when entry
+          (let [savePath (str out-dir File/separatorChar (.getName entry))
+                saveFile (File. savePath)]
+            (if (.isDirectory entry)
+              (if-not (.exists saveFile)
+                (.mkdirs saveFile))
+              (let [parentDir (File.
+                                (.substring
+                                  savePath
+                                  0
+                                  (.lastIndexOf savePath (int File/separatorChar))))]
+                (if-not (.exists parentDir)
+                  (.mkdirs parentDir))
+                (clojure.java.io/copy stream saveFile)))
+            (recur (.getNextEntry stream))))))))
+
+(defn fetch-resource
+  "Downloads a resource(zip file) and expands it to a tmp dir.
+  Returns the absolute path of the expansion dir."
+  [url]
+  (f/attempt-all [creation-args (into-array FileAttribute [])
+                  out-dir       (Files/createTempDirectory "out" creation-args)
+                  zip-stream    (-> @(http/get url)
+                                    :body
+                                    (ZipInputStream.))
+                  _             (extract-zip! zip-stream out-dir)]
+
+    (str out-dir)
+    (f/when-failed [err] err)))
+
+(defn initial-container-of
+  "Takes a path to a directory, name and image and builds the initial image.
+   This image is used by Bob as the starting image which holds the initial
+   state for the rest of the steps. Works by copying the contents to the
+   container and returns the id of the container. Deletes the temp folder
+   after it."
+  [^String path ^String image]
+  (f/attempt-all [img (pull image)
+                  id  (build img nil nil)
```

Comment:
I do have similar suggestion here as in the `commit-to-image`

In addition, it looks like the `build` function expects `^List` for second and third arguments. Is it fine that these arguments are specified as `nil` instead of empty list?